Add c10::irange to ExecuTorch

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #8554

irange is a header-only utility that is both more readable than the usual way to write for loops and also solves -Wsign-compare issues. I've previously vetted that it generates the same assembly as a regular for loop.

Differential Revision: [D69817195](https://our.internmc.facebook.com/intern/diff/D69817195/)

Commit:
Add c10::irange to ExecuTorch

irange is a header-only utility that is both more readable than the usual way to write for loops and also solves -Wsign-compare issues. I've previously vetted that it generates the same assembly as a regular for loop.

Differential Revision: [D69817195](https://our.internmc.facebook.com/intern/diff/D69817195/)

[ghstack-poisoned]

kimishpatel commented on runtime/core/portable_type/c10/c10/targets.bzl:
```diff
@@ -26,6 +26,7 @@ def define_common_targets():
             "util/TypeSafeSignMath.h",
             "util/bit_cast.h",
             "util/floating_point_utils.h",
+            "util/irange.h",
```
is irange.h not part of headers from pytorch? or we decided that is not worth sharing?

swolchok replied:
it is a pytorch header, but it is used in executorch core. therefore it must be copied over just like everything else in runtime/core/portable_type/c10/c10 to keep the build hermetic for embedded users

kimishpatel commented on runtime/core/portable_type/c10/c10/util/irange.h:
```diff
@@ -0,0 +1,123 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#pragma once
```
If this file is copied over, can you add a comment here to the effect?

swolchok replied:
> can you add a comment

no. the file should be identical to the version in pytorch core.

mergennachin commented on runtime/core/portable_type/c10/c10/util/irange.h:
```diff
@@ -0,0 +1,123 @@
+
+#include <c10/util/TypeSafeSignMath.h>
+
+#include <algorithm>
```
does it need `<algorithm>`?

swolchok replied:
not relevant; this is a direct copy from pytorch core.

mergennachin commented on runtime/core/portable_type/tensor_impl.cpp:
```diff
@@ -11,6 +11,8 @@
 #include <algorithm>
```
not related to this diff per se, but makes me curious why we need <algorithm> in this file too.

swolchok replied:
[std::equal](https://en.cppreference.com/w/cpp/algorithm/equal)

mergennachin commented on runtime/core/portable_type/c10/c10/util/irange.h:
```diff
@@ -0,0 +1,123 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
```
+1

Add BSD license header

Also, i have a task to add a lint rule: https://github.com/pytorch/executorch/issues/8418

Commit:
Update on "Add c10::irange to ExecuTorch"


irange is a header-only utility that is both more readable than the usual way to write for loops and also solves -Wsign-compare issues. I've previously vetted that it generates the same assembly as a regular for loop.

Differential Revision: [D69817195](https://our.internmc.facebook.com/intern/diff/D69817195/)

[ghstack-poisoned]

Commit:
Update on "Add c10::irange to ExecuTorch"


irange is a header-only utility that is both more readable than the usual way to write for loops and also solves -Wsign-compare issues. I've previously vetted that it generates the same assembly as a regular for loop.

Differential Revision: [D69817195](https://our.internmc.facebook.com/intern/diff/D69817195/)

[ghstack-poisoned]
